app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,109 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import mozilla.components.support.utils.SafeIntent
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Source
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.WARM
+
+/**
+ * Tracks application startup source, type, and whether or not activity has savedInstance to restore
+ * the activity from. Sample metric = [source = COLD, type = APP_ICON, hasSavedInstance = false].
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppStartupSourceAndType]and finally report them on Activity's onResume() function.
+ * If application wasn't backgrounded before onResume() is called, we don't report the metric.
+ */
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var appOpenedAllSourceStartup = AppStartupSourceAndType()
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
+        setAppStartupType()
+        setStartupSourceFromIntent(safeIntent, false)
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
+        setAppStartupType()
+        setStartupSourceFromIntent(safeIntent, true)
+    }
+
+    fun onHomeActivityOnRestart() {
+        appOpenedAllSourceStartup.type = Type.HOT
+        appOpenedAllSourceStartup.hasSavedInstance = false
+    }
+
+    fun receivedIntentInHomeActivity(safeIntent: SafeIntent) {
+        setStartupSourceFromIntent(safeIntent, false)
+    }
+
+    private fun setAppStartupType() {
+        appOpenedAllSourceStartup.type = if (wasAppCreateCalledBeforeActivityCreate) COLD else WARM
+        wasAppCreateCalledBeforeActivityCreate = false
```

Comment:
nit; Ideally this is called directly in `onCreate` or similar; otherwise someone else could call this because it seems convenient to call and it might break our boolean flag (note: boolean flags are also fragile (like this) and thus a red flag; one way around them are state machines but they add a lot of complexity and usually aren't worth it for simple cases).

Reply:
Done! initially, I was trying to remove code duplicate but this makes sense.